project-koku/koku
# COST-2066: Access denied errors during AWS org unit crawling (part one)

Fixes [COST-2066](https://issues.redhat.com/browse/COST-2066)

Changes proposed in this PR:
* We were causing a lot of access denied errors with AWS org unit crawler. Now we check the permissions, and if we do not have access we store the failure in the newly added additional context section of the provider to ensure that we don't keep causing access denied errors. 
*

Testing Instructions:
1. Add a real AWS source with org units to ensure the org unit functionality still works. 
- When the source is added it won't trigger the crawler, the crawler only triggers during the cron job or our internal masu endpoint. Therefore, you will need to hit the internal masu endpoint to run the crawler:
```
http://localhost:5042/api/cost-management/v1/crawl_account_hierarchy/?provider_uuid=e87f3dfe-9f7b-4039-9ad9-baeb5de1febb
```

2. Second update your real aws source to deny the aws org unit function, or if you are lazy like me you can add the line below  in the 137 - 139 lines of the `aws_org_unit_crawler.py ` file. 
```
raise ClientError(operation_name="", error_response={})
```
3. Rerun the crawler and check the logs for the following:
```
koku-worker_1     | [2022-02-08 01:44:15,214] INFO e4f2d584-430a-4b26-8ca3-2e0892353c34 Incomplete aws org_unit permissions for provider_uuid: e87f3dfe-9f7b-4039-9ad9-baeb5de1febb and account_id: 226558605289
```
4. Check the database:
Don't forget this table is in the public schema. 
```
postgres=# SELECT * FROM api_provider;
-[ RECORD 1 ]----------+-------------------------------------
uuid                   | e87f3dfe-9f7b-4039-9ad9-baeb5de1febb
name                   | QE AWS
type                   | AWS
setup_complete         | t
created_timestamp      | 2022-02-08 01:10:29.900052+00
data_updated_timestamp | 2022-02-08 01:11:30.284397+00
active                 | t
authentication_id      | 1
billing_source_id      | 1
created_by_id          | 1
customer_id            | 1
infrastructure_id      |
paused                 | f
additional_context     | {"crawl_hierarchy": false}
```
5. Check the sources endpoint:
```
{
            "id": 1,
            "uuid": "e87f3dfe-9f7b-4039-9ad9-baeb5de1febb",
            "name": "QE AWS",
            "source_type": "AWS",
            "authentication": {
                "role_arn": "",
                "credentials": {
                    "role_arn": ""
                }
            },
            "billing_source": {
                "bucket": "",
                "data_source": {
                    "bucket": ""
                }
            },
            "provider_linked": true,
            "active": true,
            "paused": false,
            "current_month_data": true,
            "previous_month_data": true,
            "has_data": true,
            "infrastructure": {},
            "cost_models": [],
            "additional_context": {
                "crawl_hierarchy": false
            }
        }
```

- [x]  I still need to update the unittests.
